File: Offer 28. 对称的二叉树.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }

        return compare(root.left, root.right);
    }

    boolean compare(TreeNode node1, TreeNode node2) {
        if (node1 == null && node2 == null) {
            return true;
        }

        if (node1 == null || node2 == null) {
            return false;
        }

        // 两个节点的各自的左节点和另一节点的右节点进行比较
        return (node1.val == node2.val) && compare(node1.left, node2.right) && compare(node1.right, node2.left);
    }
}

File: Offer 28. 对称的二叉树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# 递归
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        def recur(A, B):
            if not A and not B:
                return True
            elif not A or not B or A.val != B.val:
                return False
            return recur(A.left, B.right) and recur(A.right, B.left)

        return recur(root.left, root.right) if root else True

# 迭代
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
            
        q = collections.deque([root.left, root.right])
        while q:
            # 两两比较
            ln, rn = q.popleft(), q.popleft()
            if not ln and not rn:
                continue
            elif not ln or not rn or ln.val != rn.val:
                return False
            
            # 注意进队顺序
            q.append(ln.left)
            q.append(rn.right)

            q.append(ln.right)
            q.append(rn.left)

        return True